### ISSUE TITLE
NetworkManager updateNode notifies listeners in a wrong way

### ISSUE THREAD

**Issue by triller-telekom** [zsmartsystems/com.zsmartsystems.zigbee#649]
I am currently debugging why openHAB as a client of this library starts a discovery for a node twice.

I dug down the code and stumbled across the `updateNode()` method of the `ZigbeeNetworkManager`:

https://github.com/zsmartsystems/com.zsmartsystems.zigbee/blob/36f8412da5f6e2f812cd42a7f0b02768772de427/com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java#L1369-L1421

The decision of whether to inform the listeners about an _added_ or _updated_  node is taken in line 1398. However this decision is based on the fact that the discovery for that particular node has been finished or not.

So whenever a node gets changed and its discovery has not been completed yet (this is a potentially long running thread in the background), this will result in a call to `listener.nodeAdded()` instead of `listener.nodeUpdated()`.

IMHO the decision about updated vs added should only be because of changed properties in the node itself.

In the case of openHAB, there a multiple discoveries started because it acts as a `nodeAdded()` listener and whenever a new node occurs, it starts to read out its zigbee features. if the above decision would be corrected and `nodeUpdated()` would be called instead of `nodeAdded()`, this issue would be resolved.

**Comment by cdjackson** [comment]
Thanks Stefan. IMHO we should probably have the following -:

1. On first discovery, send the ```nodeAdded``` notification.
2. On subsequent updates before discovery is complete, we probably should not sent any notification.
3. Once discovery is completed, then we should send the second ```nodeAdded``` notification.
4. Subsequent updates should send the ```nodeUpdated``` notification.

ie - number 2 in this list would need to be added as it is where the second ```nodeAdded``` notification is coming from. I've not checked, but I guess the discovery service is picking up on this and adding the second discovery thread.

> IMHO the decision about updated vs added should only be because of changed properties in the node itself.

Can you explain what you mean by this exactly?

> In the case of openHAB, there a multiple discoveries started because it acts as a nodeAdded() listener and whenever a new node occurs, it starts to read out its zigbee features.

It shouldn't. openHAB should not do this unless the discovery is complete (by which it uses the ```isDiscoveryComplete``` method in the node which checks that the node has received both the node and simple descriptors. This I think is totally unrelated to having two discovery threads running as that is in the ZSS libs.

**Comment by triller-telekom** [comment]
I agree that we have to do something about number 2. Whether we should suppress notifications or rather send an `updateNode`, I don't know. IMHO the node gets updated during discovery and there might be consumers interested in this intermediate information, so I would expect an `updateNode`.

Why do you want to send an **additional** `nodeAdded` in point 3? A consumer would then receive **two** `nodeAdded` events for a single device which was added. As stated above, I would also rather expect a `nodeUpdated` here. IMHO we should only have **one** added event per new device.

>> IMHO the decision about updated vs added should only be because of changed properties in the node itself.

> Can you explain what you mean by this exactly?

I mean that in the code the variable `updated` is calculated only from the fact that it belongs to the `nodeDiscoveryComplete` set, i.e. `updated = nodeDiscoveryComplete.contains(node.getIeeeAddress())`.

I would have expected that for making the decision that a node was updated, we would check ALL properties of a `ZigbeeNode`, i.e. `nodeState`, `ieeeAddress`, `endPoints`, etc. and compare them against the existing node. So something like `updated = ! node.equals(existingNode)`.

> It shouldn't. openHAB should not do this unless the discovery is complete (by which it uses the isDiscoveryComplete method in the node which checks that the node has received both the node and simple descriptors

Well, this is what I get:

```
2019-06-03T12:31:10.938+0200   433  DEBUG          [symbolicName=org.openhab.binding.zigbee, className=org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - Stefan: start discovery because node was added: node=ZigBeeNode [state=UNKNOWN, IEEE=000D6F000BBD2B5B, NWK=2DC2, Type=UNKNOWN]
2019-06-03T12:31:10.939+0200   433  DEBUG          [symbolicName=org.openhab.binding.zigbee, className=org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - Stefan: nodeDiscovered(): node=ZigBeeNode [state=UNKNOWN, IEEE=000D6F000BBD2B5B, NWK=2DC2, Type=UNKNOWN]
2019-06-03T12:31:11.359+0200   433  DEBUG          [symbolicName=org.openhab.binding.zigbee, className=org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - Stefan: start discovery because node was added: node=ZigBeeNode [state=ONLINE, IEEE=000D6F000BBD2B5B, NWK=2DC2, Type=UNKNOWN]
2019-06-03T12:31:11.359+0200   433  DEBUG          [symbolicName=org.openhab.binding.zigbee, className=org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - Stefan: nodeDiscovered(): node=ZigBeeNode [state=ONLINE, IEEE=000D6F000BBD2B5B, NWK=2DC2, Type=UNKNOWN]
2019-06-03T12:31:15.263+0200   433  DEBUG          [symbolicName=org.openhab.binding.zigbee, className=org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - Stefan: start discovery because node was added: node=ZigBeeNode [state=UNKNOWN, IEEE=000D6F000F24146C, NWK=0000, Type=COORDINATOR]
2019-06-03T12:31:17.923+0200   433  DEBUG          [symbolicName=org.openhab.binding.zigbee, className=org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - Stefan: start discovery because node was added: node=ZigBeeNode [state=ONLINE, IEEE=000D6F000BBD2B5B, NWK=2DC2, Type=ROUTER]
2019-06-03T12:31:17.937+0200   433  DEBUG          [symbolicName=org.openhab.binding.zigbee, className=org.openhab.binding.zigbee.discovery.ZigBeeDiscoveryService] - Stefan: nodeDiscovered(): node=ZigBeeNode [state=ONLINE, IEEE=000D6F000BBD2B5B, NWK=2DC2, Type=ROUTER]
```

As you can see the `nodeDiscovered()` method from openHAB gets a node with `Type=UNKNOWN` **twice** and thus the `isDiscovered()` check is skipped. Only on the third call, there is `Type=ROUTER` and thus the check prevents it from starting a third discovery.

**Comment by cdjackson** [comment]
> Why do you want to send an additional nodeAdded in point 3?

Some systems expect this - eg openHAB. The current functionality is to send a Node Added when the device is first added and before discovery starts, and then another Node Added when discovery completes. The openHAB binding differentiates with the ```isDiscoveryComplete``` method. We can of course change this, but I think the binding will stop working (although I've not checked this).

**Comment by cdjackson** [comment]
@triller-telekom on further thought, I think we need to implement this as I originally suggested -:

1. On first discovery, send the nodeAdded notification.
1. On subsequent updates before discovery is complete, we probably should not sent any notification.
1. Once discovery is completed, then we should send the second nodeAdded notification.
1. Subsequent updates should send the nodeUpdated notification.

My rational here is that if we change to send the ```nodeAdded``` only on the initial discovery , and before any services (is endpoints/clusters) are known, then in order to find when a device has completed discovery, an app now has to use the ```nodeUpdated``` listener. This may be called quite often, and it requires further implementation in order to differentiate from an "end of discovery" ```nodeUpdated``` or just a LQI/neighbors etc ```nodeUpdated```. I think it is therefore better to send a ```nodeAdded``` when discovery completes as we are doing now.

WDYT?

**Comment by triller-telekom** [comment]
So a client would then ALWAYS get 2 `nodeAdded` events for the very same node, hmm.

Your argument for doing that is that its easier for a client to distinguish a "regular" `nodeUpdated` from the end of a discovery. However, one could use the same argument for what you suggest now:

Imagine a client "receiving" the first `nodeAdded` vs the second `nodeAdded`, now he also have to match them somehow to know its the very same node and that the second call means the discovery has been completed.

So in either case its the client who has to "figure out" something. What about if we would extend the API and add a `discoveryComplete` method and then use `nodeAdded` and `nodedUpdated` only if a node was first seen or some properties of a node have changed?

**Comment by cdjackson** [comment]
Extending the API is also breaking - the same as we have done with removing the second nodeAdded notification. This is something I was trying to avoid (although currently it is now changed).

openHAB, and other implementations already implement this (this being the 2 nodeAdded calls). The library has always done this and they differentiate the two notifications with the isDiscoveryComplete method. The point is that there are only 2 nodeAdded notifications, so if a user is waiting for the device to be added and discovery complete, then they will only ever have 2 calls (Again - this is how it is currently implemented, except for this issue where it could be 3 and that causes problems!). This is more efficient than having to implement the nodeUpdated call as this will be called hundreds of times (ie after every mesh update!) which means the application is being called a LOT, and if it only cares about nodes being added, then it is very inefficient.

I do like the idea of the discoveryComplete method (or similar name maybe) - if we want to break the API, then maybe this is a clearer way of doing it. WDYT?

**Comment by triller-telekom** [comment]
What we now do is implicit:

> The library has always done this and they differentiate the two notifications with the isDiscoveryComplete method

And since the discovery mechanism is now broken (at least in the case one uses the openHAB binding with the latest ZSS master), I would rather vote for doing it explicit, so adding a `discoveryComplete` method.

Each client then knows that there is a new node via *ONE* `nodeAdded` and then once the discovery completes via `discoveryComplete` so there is no need to remember calling `isDiscoveryComplete` within `nodeAdded`.

However, since you broke the API with the 1.2 release I don't know if you like to break it again with a follow up 1.3 now...

**Comment by triller-telekom** [comment]
How do we want to proceed here?

Since https://github.com/zsmartsystems/com.zsmartsystems.zigbee/pull/671 has been merged, all users of ZSS have to implemented `ZigBeeNetworkNodeListener#nodeUpdated(ZigBeeNode node)` as you mentioned earlier:

> an app now has to use the `nodeUpdated` listener. This may be called quite often, and it requires further implementation in order to differentiate from an "end of discovery" `nodeUpdated` or just a LQI/neighbors etc `nodeUpdated`.

Have you thought about introducing a ``ZigBeeNetworkNodeListener#discoveryComplete(ZigBeeNode node)` method and breaking the API (again)? IMHO it makes sense to do so.

**Comment by triller-telekom** [comment]
Thinking about this again: The way the implementation is done in #671 all users of this library run into the problem that they do not get notified once the discovery is complete, since the second `nodeAdded` is not send anymore.

Do you know a better solution than introducing `ZigBeeNetworkNodeListener#discoveryComplete(ZigBeeNode node)`, or shall we go forward and add it?

**Comment by cdjackson** [comment]
> all users of this library run into the problem that they do not get notified once the discovery is complete, since the second nodeAdded is not send anymore.

Yes, this is as we discussed - with this implementation users need to use the ```nodeUpdated``` method.

> Do you know a better solution than introducing ZigBeeNetworkNodeListener#discoveryComplete(ZigBeeNode node), or shall we go forward and add it?

I have thought about this. At the moment, the current released code (```1.2.0```) is still the same as ```1.1.x``` and I won't release another version until this is resolved.

My current thinking is to add a ```nodeDiscovery``` method (or something similar), but this does need some thought as there are different discovery systems in use (eg there is a different system in place for ```Smart Energy``` than the standard discovery service). If we implement this, I want to ensure that it's consistent, and it will probably still be able to be called multiple times if the discovery information changes.

**Comment by triller-telekom** [comment]
> with this implementation users need to use the `nodeUpdated` method.

But you said that this method is called a lot and users cannot decide if the node was just updated or the discovery was just completed. If we would adapt openHAB for example, this would mean that on each `nodeUpdated` we would create a discovery result. This would then get checked against the current things and then removed because the thing already exists. I wonder if its worth to implement it this way to get the openHAB binding compatible with the latest zss master.

Or is your plan to make the openHAB binding only compatible with zss-1.2.0?

**Comment by cdjackson** [comment]
> But you said that this method is called a lot and users cannot decide if the node was just updated or the discovery was just completed.

Yes, and no. Yes, it's called a lot, which is why I don't like it so much, but it's perfectly possible for users to decide if discovery is complete.

> If we would adapt openHAB for example, this would mean that on each nodeUpdated we would create a discovery result.

Why would you implement that? That would not be a particularly smart thing to do I think? I don't see that it should really be any different than now. The OH service discoverer (I forget the exact class name now) should only run when discovery is complete.

> Or is your plan to make the openHAB binding only compatible with zss-1.2.0?

As above, I plan to change this, but I've only had 1 day to look at it, and I was unwell yesterday. I'm also working on the SPI work for DT which must be completed very soon now so it has to take priority.

**Comment by triller-telekom** [comment]
>> If we would adapt openHAB for example, this would mean that on each nodeUpdated we would create a discovery result.

>Why would you implement that? That would not be a particularly smart thing to do I think? I don't see that it should really be any different than now. The OH service discoverer (I forget the exact class name now) should only run when discovery is complete.

Well as you said, it has to be different than now... Within `nodeUpdated` we can call `node.isComplete`, before we create a discovery result (I forgot to mention that), however once the node is complete and is updated *afterwards*, then there would be a new discovery result created, as the node is still complete (maybe it got more neighbours for example).

**Comment by cdjackson** [comment]
>  however once the node is complete and is updated afterwards, then there would be a new discovery result created,

Please can you clearly define what the discovery result is? Is this the inbox result in OH, or something in the ZSS libraries? Or something else? Sorry, but it's not clear what you're getting at.

**Comment by triller-telekom** [comment]
Oh right, sorry if i mixed up the terms here.

With discovery result I meant the OH inbox entry. So what I am trying to say is that at some point ZSS has finished its node discovery and `node.isComplete()` will then always return true. If ZSS adds a new neighbour to the node for example, this would lead to a `nodeUpdated()` call and the OH binding will then create that inbox entry "again", but the OH framework will notice that such a `Thing` already exists and remove the entry from the inbox again.

**Comment by cdjackson** [comment]
Ok, but this sort of issue in OH could easily be resolved in OH with a simple flag.

In any case, as I said above I'm looking at adding the ```nodeDiscovered``` callback so that should resolve this.

**Comment by cdjackson** [comment]
Ok, here's what I will do...

In PR #685 I restore the original functionality of the ```nodeAdded``` listener, so there will again be 2 calls. It also withholds notifications after the initial ```nodeAdded``` until discovery is complete to avoid the problem that was originally reported in this issue.

With this change, we should have the implementation discussed [above](https://github.com/zsmartsystems/com.zsmartsystems.zigbee/issues/649#issuecomment-498742575). This should solve this issue, and avoid a breaking change in ```1.2.x```.

I decided to go down this route rather than to jump in and make a change to add another method in the  ```ZigBeeNetworkNodeListener``` interface without thinking this through. I might change the way this works in a different way for ```1.3.0``` - possibly simplifying the interface to have a single method and an enumeration to define the reason for change. It may also make better sense for the ```nodeAdded``` call to be a general call, while a ```nodeChanged``` call possibly should be registered against the ```ZigBeeNode``` rather than the ```ZigBeeNetworkManager``` which I think would be cleaner - again, this is why I've decided not to rush in to making a change which adds more methods to this interface.

I'm happy to hear your thoughts of course @triller-telekom :)
